src/cargo/core/global_cache_tracker.rs

```diff
@@ -0,0 +1,1754 @@
+//! Support for tracking the last time files were used to assist with cleaning
+//! up those files if they haven't been used in a while.
+//!
+//! Tracking of cache files is stored in a sqlite database which contains a
+//! timestamp of the last time the file was used, as well as the size of the
+//! file.
+//!
+//! While cargo is running, when it detects a use of a cache file, it adds a
+//! timestamp to [`DeferredGlobalLastUse`]. This batches up a set of changes
+//! that are then flushed to the database all at once (via
+//! [`DeferredGlobalLastUse::save`]). Ideally saving would only be done once
+//! for performance reasons, but that is not really possible due to the way
+//! cargo works, since there are different ways cargo can be used (like `cargo
+//! generate-lockfile`, `cargo fetch`, and `cargo build` are all very
+//! different ways the code is used).
+//!
+//! All of the database interaction is done through the [`GlobalCacheTracker`]
+//! type.
+//!
+//! There is a single global [`GlobalCacheTracker`] and
+//! [`DeferredGlobalLastUse`] stored in [`Config`].
+//!
+//! ## Automatic gc
+//!
+//! Some commands (primarily the build commands) will trigger an automatic
+//! deletion of files that haven't been used in a while. The interface for
+//! this is in the [`crate::core::gc`] module. The database tracks the last
+//! time an automatic gc was performed so that it is only done once per day
+//! for performance reasons.
+//!
+//! ## Manual gc
+//!
+//! The user can perform a manual garbage collection with the `cargo clean`
+//! command. That command has a variety of options to specify what to delete.
+//! Manual gc supports deleting based on age or size or both.
+//!
+//! ## Locking
+//!
+//! Usage of the database requires that the package cache is locked to prevent
+//! concurrent access. Although sqlite has built-in locking support, we want
+//! to use cargo's locking so that the "Blocking" message gets displayed, and
+//! so that locks can block indefinitely for long-running build commands.
+//! [`rusqlite`] has a default timeout of 5 seconds, though that is
+//! configurable.
```

Comment:
> so that locks can block indefinitely for long-running build commands.

This had me worried for a second that a `cargo build` will block on another `cargo build` due to gc but this is only true for manual gc and not auto-gc as we skip auto-gc if there is contention for the lock.

I don't know if we should go into that level of detail but I feel like some clarification is needed or else this might continue to raise eyebrows (from future me or others) who see this and get concerned.

Reply:
I added some more information about the opportunistic behavior of automatic gc.